File: tests/Adapter/MrCheckoutTest.php
<?php

declare(strict_types=1);

namespace Upkeep\Tests\Adapter;

use PHPUnit\Framework\Attributes\CoversClass;
use PHPUnit\Framework\TestCase;
use Upkeep\Adapter\MrCheckout;

#[CoversClass(MrCheckout::class)]
final class MrCheckoutTest extends TestCase
{
    public function testFetchRefspecCarriesWhicheverRefWasChosen(): void
    {
        // Force-refspec (+...) so re-applying an MR after it moved updates the
        // local branch instead of failing non-fast-forward. The merge ref moves
        // for a second reason the head ref does not: GitLab recomputes it when
        // the *target* gains a commit.
        self::assertSame(
            '+refs/merge-requests/2/merge:mr-2',
            MrCheckout::fetchRefspec(MrCheckout::mergeRef(2), 2),
        );
        self::assertSame(
            '+refs/merge-requests/147/head:mr-147',
            MrCheckout::fetchRefspec(MrCheckout::headRef(147), 147),
        );
    }

    /**
     * The merge ref wins, and that is the correction.
     *
     * `/head` is the contributor's branch; `/merge` is that branch merged into
     * the *current* tip of the target, which is what CI analyses. Measured on
     * pathauto: 23 of 25 open merge requests have a merge tree that differs
     * from their head tree, with branches 7 to 41 commits behind. Checking the
     * head meant agreeing with CI by luck.
     */
    public function testTheMergeRefIsPreferredOverTheHeadRef(): void
    {
        $both = ['refs/merge-requests/4/head', 'refs/merge-requests/4/merge'];

        self::assertSame('refs/merge-requests/4/merge', MrCheckout::preferredRef($both, 4));
        self::assertSame('refs/merge-requests/4/merge', MrCheckout::preferredRef(array_reverse($both), 4));
    }

    /**
     * No merge ref means GitLab could not merge the branch into its target —
     * normally a conflict. The branch is still checkable, but the fallback is
     * announced, because a branch-only verdict is not the one CI would give.
     */
    public function testOnlyAHeadRefFallsBackToItAndIsWorthSaying(): void
    {
        self::assertSame(
            'refs/merge-requests/4/head',
            MrCheckout::preferredRef(['refs/merge-requests/4/head'], 4),
        );

        $warning = MrCheckout::noMergeRefWarning(4);
        self::assertStringContainsString('conflict', $warning);
        self::assertStringContainsString('not what CI runs', $warning);
    }

    /** Another merge request's refs are not this one's. */
    public function testRefsBelongingToADifferentMergeRequestAreNotAccepted(): void
    {
        self::assertNull(MrCheckout::preferredRef(
            ['refs/merge-requests/5/merge', 'refs/merge-requests/5/head'],
            4,
        ));
        self::assertNull(MrCheckout::preferredRef([], 4));
    }

    public function testNoRefsAtAllIsARefusalNamingBoth(): void
    {
        $refusal = MrCheckout::noRefsAtAll(4);

        self::assertStringContainsString('refs/merge-requests/4/merge', $refusal->getMessage());
        self::assertStringContainsString('refs/merge-requests/4/head', $refusal->getMessage());
    }

    public function testBranchNameIsDerivedFromTheIid(): void
    {
        self::assertSame('mr-2', MrCheckout::branchName(2));
    }

    /**
     * The record answers only what the working copy cannot.
     *
     * On a managed branch there is no other way to know what it was cut from,
     * so the record wins. On a real branch, that branch *is* the base and the
     * record is at best a description of some earlier state.
     */
    public function testTheRecordedBaseIsUsedOnlyWhenTheBranchCannotAnswer(): void
    {
        self::assertSame('2.0.x', MrCheckout::resolveBaseBranch('mr-2', '2.0.x'), 'managed: ask the record');
        self::assertSame('2.0.x', MrCheckout::resolveBaseBranch('patch-3559057', '2.0.x'));
        self::assertSame('2.0.x', MrCheckout::resolveBaseBranch(null, '2.0.x'), 'detached: ask the record');
    }

    /**
     * This assertion used to be the other way round, and it was wrong.
     *
     * A stale record was permanent once written: check a patch, which records
     * the issue's base as 1.0.x, then move the working copy to the branch a
     * merge request targets, and the check still refused on a 1.0.x base —
     * advising `upkeep dev <module> --branch=2.0.x`, which is exactly what had
     * just been run. The nightly full check hit it twice.
     *
     * Clearing the record on checkout would not fix it. The module working
     * copy is a real clone and `git checkout` in it is ordinary use, so the
     * record can go stale without upkeep ever being told; the reader has to be
     * right rather than the writer being diligent.
     */
    public function testARealBranchOutranksAStaleRecord(): void
    {
        self::assertSame('2.0.x', MrCheckout::resolveBaseBranch('2.0.x', '1.0.x'));
    }

    public function testCurrentBranchIsTheBaseWhenNothingIsRecorded(): void
    {
        self::assertSame('1.0.x', MrCheckout::resolveBaseBranch('1.0.x', null));
    }

    public function testBaseCannotBeResolvedFromAnMrBranchWithoutARecord(): void
    {
        $this->expectException(\Upkeep\Adapter\AdapterException::class);
        $this->expectExceptionMessageMatches('/mr-7/');
        MrCheckout::resolveBaseBranch('mr-7', null);
    }

    public function testBaseCannotBeResolvedFromADetachedHeadWithoutARecord(): void
    {
        $this->expectException(\Upkeep\Adapter\AdapterException::class);
        MrCheckout::resolveBaseBranch(null, null);
    }

    public function testNativeBaseValidationPassesWhenTargetMatchesBase(): void
    {
        MrCheckout::assertNativeBase($this->mergeRequest(targetBranch: '2.0.x'), '2.0.x', 'widget');
        $this->addToAssertionCount(1);
    }

    /**
     * The refusal names the command that fixes it.
     *
     * Not a nicety: a module whose default branch is not the branch its merge
     * requests target hits this on the very first `check`, and the nightly
     * full check hit it against a real module — field_visibility_conditions
     * clones on 1.0.x while both of its open MRs target 2.0.x. "Apply the MR
     * in an environment whose base is its target branch" is true and leaves
     * you to work out how; the how is one command, and every other refusal in
     * this tool ends in something you can paste.
     */
    public function testBackportIsRejectedAndSaysHowToGetTheRightBase(): void
    {
        // The only real-world shape: an MR targeting a branch the environment
        // does not have checked out as its base (e.g. fvc !2 targets 2.0.x,
        // environment base is 1.0.x) — out-of-scope backport testing.
        $this->expectException(\Upkeep\Adapter\AdapterException::class);
        $this->expectExceptionMessageMatches('/MR !2 targets branch "2\.0\.x".*branch "1\.0\.x".*[Bb]ackport/s');
        $this->expectExceptionMessageMatches('/upkeep dev widget --branch=2\.0\.x/');
        MrCheckout::assertNativeBase($this->mergeRequest(targetBranch: '2.0.x'), '1.0.x', 'widget');
    }

    private function mergeRequest(string $targetBranch): \Upkeep\Gitlab\MergeRequest
    {
        return new \Upkeep\Gitlab\MergeRequest(
            iid: 2,
            title: 'Draft: Automated Project Update Bot fixes',
            state: 'opened',
            authorUsername: 'Project-Update-Bot',
            authorId: 66574,
            sourceBranch: 'project-update-bot-only',
            targetBranch: $targetBranch,
            draft: true,
            detailedMergeStatus: 'draft_status',
            headSha: '6e77deef6897f4ac6549d26e66fe0ab3d300bf76',
            webUrl: 'https://git.drupalcode.org/project/field_visibility_conditions/-/merge_requests/2',
        );
    }
}

File: go/internal/adapter/mrcheckout_test.go
package adapter

import (
	"strings"
	"testing"

	"github.com/owenbush/upkeep/internal/gitlab"
)

// CI analyses the merge, not the branch. Measured on pathauto: of 25 open
// merge requests, 23 have a merge tree that differs from their head tree.
func TestTheMergeRefIsPreferredOverTheHead(t *testing.T) {
	both := []string{
		"refs/heads/2.0.x",
		HeadRef(12),
		MergeRef(12),
	}

	ref, found := PreferredRef(both, 12)
	if !found {
		t.Fatal("neither ref was found")
	}
	if ref != MergeRef(12) {
		t.Errorf("chose %q, want the merge ref", ref)
	}
}

// GitLab computes no merge ref for a merge request that conflicts with its
// target, so the fallback is a diagnosis rather than a detail.
func TestWithNoMergeRefTheHeadIsUsedLoudly(t *testing.T) {
	ref, found := PreferredRef([]string{HeadRef(12)}, 12)
	if !found || ref != HeadRef(12) {
		t.Fatalf("got %q, %v", ref, found)
	}

	warning := NoMergeRefWarning(12)
	for _, want := range []string{"!12", "conflict", "not what CI runs", "says nothing"} {
		if !strings.Contains(warning, want) {
			t.Errorf("the warning does not carry %q:\n%s", want, warning)
		}
	}
}

// Neither ref at all is not a state to guess about: the iid is wrong, or the
// merge request was removed.
func TestNeitherRefIsARefusalNotAFallback(t *testing.T) {
	if ref, found := PreferredRef([]string{"refs/heads/2.0.x", MergeRef(99)}, 12); found {
		t.Errorf("found %q for a merge request that publishes nothing", ref)
	}

	err := ErrNoRefsAtAll(12)
	for _, want := range []string{"!12", "refs/merge-requests/12/merge", "Check the merge request number"} {
		if !strings.Contains(err.Error(), want) {
			t.Errorf("the refusal does not carry %q: %v", want, err)
		}
	}
}

// Force-updating, because the merge ref moves whenever the *target* gains a
// commit — a re-check after an unchanged merge request can still be a
// different tree.
func TestTheRefspecForcesAndLandsOnTheManagedBranch(t *testing.T) {
	spec := FetchRefspec(MergeRef(12), 12)

	if !strings.HasPrefix(spec, "+") {
		t.Errorf("refspec %q does not force-update", spec)
	}
	if !strings.HasSuffix(spec, ":"+ManagedBranchForMergeRequest(12)) {
		t.Errorf("refspec %q does not land on the managed branch", spec)
	}
}

// The record is a fallback, not an override. It used to win outright, which
// made a stale record permanent — found by the nightly full check, twice.
func TestARealBranchOutranksAStaleRecord(t *testing.T) {
	got, err := ResolveBaseBranch("2.0.x", "1.0.x")
	if err != nil {
		t.Fatalf("unexpected refusal: %v", err)
	}
	if got != "2.0.x" {
		t.Errorf("got %q, want the branch the working copy is actually on", got)
	}
}

// The record answers the one question the working copy cannot: what a managed
// branch was cut from, while you are sitting on it.
func TestTheRecordAnswersForAManagedBranch(t *testing.T) {
	for _, branch := range []string{ManagedBranchForMergeRequest(12), ManagedBranchForPatch(3603341)} {
		got, err := ResolveBaseBranch(branch, "2.0.x")
		if err != nil {
			t.Errorf("%s: unexpected refusal: %v", branch, err)

			continue
		}
		if got != "2.0.x" {
			t.Errorf("%s: got %q", branch, got)
		}
	}
}

// Cutting a disposable branch from a work branch would test the contribution
// *plus* unpushed work, and report the result as a verdict on the
// contribution alone.
func TestAWorkBranchIsNeverABase(t *testing.T) {
	_, err := ResolveBaseBranch("3603341-drupal-12-compatibility", "")
	if err == nil {
		t.Fatal("a work branch was accepted as a base")
	}
	for _, want := range []string{"your own work branch", "on top of that work", "git -C <module> checkout"} {
		if !strings.Contains(err.Error(), want) {
			t.Errorf("the refusal does not carry %q: %v", want, err)
		}
	}

	// And a record behind it is still the answer.
	got, err := ResolveBaseBranch("3603341-drupal-12-compatibility", "2.0.x")
	if err != nil {
		t.Fatalf("unexpected refusal: %v", err)
	}
	if got != "2.0.x" {
		t.Errorf("got %q", got)
	}
}

// A detached HEAD with nothing recorded has no base to find.
func TestADetachedHeadWithNoRecordIsARefusal(t *testing.T) {
	_, err := ResolveBaseBranch("", "")
	if err == nil {
		t.Fatal("a base was resolved from nothing")
	}
	if !strings.Contains(err.Error(), "HEAD is detached") {
		t.Errorf("the refusal does not say why: %v", err)
	}

	got, err := ResolveBaseBranch("", "2.0.x")
	if err != nil {
		t.Fatalf("unexpected refusal: %v", err)
	}
	if got != "2.0.x" {
		t.Errorf("got %q", got)
	}
}

func TestAManagedBranchWithNoRecordIsARefusalThatNamesIt(t *testing.T) {
	_, err := ResolveBaseBranch(ManagedBranchForMergeRequest(12), "")
	if err == nil {
		t.Fatal("a base was resolved from a managed branch alone")
	}
	if !strings.Contains(err.Error(), "upkeep-managed branch \"mr-12\"") {
		t.Errorf("the refusal does not name the branch: %v", err)
	}
}

// A verdict from the wrong base is about code nobody proposed.
func TestAMergeRequestMayOnlyBeAppliedAgainstItsOwnTarget(t *testing.T) {
	mr := gitlab.MergeRequest{IID: 12, TargetBranch: "2.0.x"}

	if err := AssertNativeBase(mr, "2.0.x", "pathauto"); err != nil {
		t.Errorf("its own target was refused: %v", err)
	}

	err := AssertNativeBase(mr, "1.0.x", "pathauto")
	if err == nil {
		t.Fatal("backport testing was allowed")
	}
	// Named, not described: every other refusal here ends in something you can
	// paste.
	if !strings.Contains(err.Error(), "upkeep dev pathauto --branch=2.0.x") {
		t.Errorf("the refusal does not name the command that fixes it:\n%v", err)
	}
	for _, want := range []string{"!12", "\"2.0.x\"", "\"1.0.x\"", "Backport testing is out of scope"} {
		if !strings.Contains(err.Error(), want) {
			t.Errorf("the refusal does not carry %q:\n%v", want, err)
		}
	}
}
